### PR TITLE
Detect closures assigned to binding in block

### LINKED ISSUE #58497: Suboptimal error when moving a Copy type in a closure
```rust
#[derive(Copy, Clone)]
struct Bar;

fn foo(bar: Bar) -> impl FnOnce() -> Bar {
    || bar
}
```

([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=aafeeeb1c964977b6f29a02e1b4ee9a2)) main error message is:

```
error[E0373]: closure may outlive the current function, but it borrows `bar`, which is owned by the current function
 --> src/lib.rs:5:5
  |
5 |     || bar
  |     ^^ --- `bar` is borrowed here
  |     |
  |     may outlive borrowed value `bar`
  |
note: closure is returned here
```

I spent a long time trying to figure out where I was creating a reference until I finally figured out that `Copy` types are always captured by reference, even if you move them in the closure. It would be helpful if it mentioned the fact that while this looks like a move it's actually copying out of a reference because the type is `Copy`.

### PR DESCRIPTION
**Pull request by estebank**
Fix #58497.

### CODE REVIEW

File: src/tools/rustfmt/tests/target/issue_4110.rs [hunk 1]
```diff
@@ -20,6 +20,7 @@ fn bindings() {
                 category,
                 span,
                 &format!("`{}`", name),
+                "function",
```
Comment (ytmimi) [inline src/tools/rustfmt/tests/target/issue_4110.rs:23]: Was this change necessary? I'm curious why a rustfmt test needed to change.

Comment (estebank) [reply src/tools/rustfmt/tests/target/issue_4110.rs:23]: I added a new argument in rustc. The rustfmt change is just to keep compatibility.

Comment (calebcartwright) [reply src/tools/rustfmt/tests/target/issue_4110.rs:23]: @ytmimi - Technically I don't think it's necessary because we never need to compile these files, but I don't think the extra arg impacts the integrity of the rustfmt test in any way (I assumed it was simply changed based on a find+replace using the ident)

Comment (estebank) [reply src/tools/rustfmt/tests/target/issue_4110.rs:23]: Yeah, I blanket searched for all the function calls.

**Review by davidtwco — APPROVED** [review]
